wordpress-mobile/WordPress-iOS
# Media: adios, and integrate new export and thumbnail services with MediaService.

After many PRs, this finally integrates live exporting, uploading, thumbnails, etc, using the new Media services. Thus, this brings both my Media and WPiOS journey to a close... for now 😃 😭 ❤️ 🦅 
 
Huge thanks to @SergioEstevao who did all the original Media implementation by himself, which I could copy along the way 😄 .

Another huge thanks to @frosty for bearing with both all of my Media PRs, AND my ridiculous iPad PRs, over the last couple of years, 👏 

### Feature Flag
This is currently feature flagged as enabled for local and CI test builds. Once we're confident with both testing and timing, it could be bumped up for internal testing and an eventual release.

We'll really need a good bit of testing from users who actually use our editor with photos, such as our internal testers. Everything _seems_ to work very nicely, but who knows what edge cases lurk underneath. I'll be more than happy to look at new issues over the next months as they crop up, open-source ftw.

### MediaService

Our `MediaService` lives on! Although a lot lighter than it used to be. Ultimately, `MediaService` should be the ultimate coordinator of the different services, including the upcoming "upload manager/service". Once we fully switch to using only our new Media exporters, we can nuke the following code (of which I'd love to do the honors myself 😄):

- `Services/MediaService+Legacy.h/m`
- `Extensions/PHAsset+Exporters.swift`
- `Extensions/NSURL+Exporters.swift`
- `Extensions/UIImage+Exporters.swift`

### Thumbnails
A hard as I've tried, and believe me I've tried pretty hard, I can not get rid of `absoluteThumbnailLocalURL`. We're entirely dependent on it having a readily available thumbnail image to show as a placeholder image. For now, and probably for a while, we'll go ahead and honor that by pre-exporting a general placeholder thumbnail while we export Media initially.

It is highly recommended that we, over time, switch all references to `absoluteThumbnailLocalURL` to an asynchronous request for a preferably sized thumbnail image. It's just not-all-that-awesome to have `absoluteThumbnailLocalURL` running around advertising a thumbnail to the rest of the code-base. The method at `MediaService.exportPlaceholderThumbnailForMedia:...` documents some details and what's currently happening.

### Future improvements

- It would be very handy to track all occurrences of the different export errors, so we know exactly how those types of cases are occurring.
- `MediaAssetExporter` can be expanded to return an `NSProgress` instance when exporting media from the user's library. This will be particularly handy when the device would have to download asset data from iCloud, letting us show more than "preparing" for several seconds.
- The exporters themselves could also be expanded to allow cancellation of exports, such as `MediaAssetExporter`.
- `MediaThumbnailService` could probably use some unit tests, but is generally covered via the underlying `MediaImageExporter` and its tests.

### Notes

- `MediaThumbnailService.swift` has a wonky diff here as it had a couple of minor changes, but a "big" change in that I switched where the `context.performBlock` was wrapping the code.
- Whether we like it or not, using `AVAssetExportPresetPassthrough` on video exports is much faster than any other preset.
  - We are currently defaulting to `AVAssetExportPresetHighestQuality`, which can take a bit to prepare.

### Testing

1. Make sure it compiles 😄 
2. Run all projects tests.

-- -

1. Test the site's Media Library section and make sure everything loads as expected.
2. Clear the cache in App Settings, then retest to make sure it re-generates thumbnails.
3. Make sure to test photos, videos, Gifs, documents, all render (or don't) as expected.

---

1. Enable the Visual editor.
2. Test that inserting Media from the device library works as expected with JPEGs, PNGs, GIFs, videos, and anything else we should support from the device library.
3. Test that inserting Media from using the camera works as expected, with both a photo and video.
4. Test that pasting a photo into the Editor's text view works as expected.
5. Test that pasting a gif into the Editor's text view works as expected.
6.Test that selecting a photo or video from the Media Library picker works as expected.
7. Test multiple photos at the same time, with decent files for a lengthy upload time.
8. Test multiple videos at the same time, with decent files for a lengthy upload time.
9. Repeat these steps for the Aztec editor.
10. Maybe repeat these steps for the legacy editor?
11. 🎉 

---

1. Test disabled the feature altogether.
2. Return `false` in `FeatureFlag.swift` under the `. newMediaExports` case.
3. Test both the Media Library and Editors work as usual, using the legacy code.

---

Needs review: @frosty, @SergioEstevao, definitely no rush on this! I know Sergio is AFK, this can also be merged with lesser visibility behind the feature flag.

🎩 tip
